gem/optimise.py

```diff
@@ -230,13 +230,47 @@ def delta_elimination(sum_indices, factors):
     return sum_indices, [e for e in factors if e != one]
 
 
+def associate_product(factors):
+    """Apply associativity rules to construct an operation-minimal product tree.
+
+    For best performance give factors that have different set of free indices.
+    """
```

Comment:
"Create product which has best performance (i.e. lowest number of flops) from a given list of factors with (potentially) different set of free indices" ?